awc/src/middleware/redirect.rs

```diff
@@ -156,79 +161,59 @@ where
                     StatusCode::MOVED_PERMANENTLY
                     | StatusCode::FOUND
                     | StatusCode::SEE_OTHER
+                    | StatusCode::TEMPORARY_REDIRECT
+                    | StatusCode::PERMANENT_REDIRECT
                         if *max_redirect_times > 0 =>
                     {
-                        let org_uri = uri.take().unwrap();
-                        // rebuild uri from the location header value.
-                        let uri = rebuild_uri(&res, org_uri)?;
+                        let is_redirect = res.head().status == StatusCode::TEMPORARY_REDIRECT
+                            || res.head().status == StatusCode::PERMANENT_REDIRECT;
 
-                        // reset method
-                        let method = method.take().unwrap();
-                        let method = match method {
-                            Method::GET | Method::HEAD => method,
-                            _ => Method::GET,
-                        };
+                        let prev_uri = uri.take().unwrap();
+
+                        // rebuild uri from the location header value.
+                        let next_uri = build_next_uri(&res, &prev_uri)?;
 
                         // take ownership of states that could be reused
+                        let body = body.take();
                         let addr = addr.take();
                         let connector = connector.take();
-                        let mut max_redirect_times = *max_redirect_times;
-
-                        // use a new request head.
-                        let mut head = RequestHead::default();
-                        head.uri = uri.clone();
-                        head.method = method.clone();
 
-                        let head = RequestHeadType::Owned(head);
-
-                        max_redirect_times -= 1;
+                        // reset method
+                        let method = if is_redirect {
```

Comment:
maybe i'm reading this wrong (and i think there isn't yet a test for this) but the 307 and 308 status codes should preserve the method, this looks like it's flipped the logic

Reply:
Added some more test cases 👍🏼